frontend/src/main/scala/bloop/cli/Proxy.scala

```diff
@@ -0,0 +1,90 @@
+    * If environement variable are not found in the state, the property corresponding is cleared,
+    * to let user add and remove env var.
+    * */
+  def setProxyJvmSettings(state : State) : Unit = {
+    //no_proxy environment variable seems not be binded to a protocol.
+
+    val maybeNoProxy = Option(state.commonOptions.env.getProperty("no_proxy")).map(_.replace(',','|'))
+
+    val maybeHttpProxy = state.commonOptions.env.getProperty("http_proxy")
+    if(maybeHttpProxy != null) {
+      try {
+        val url =  new URL(maybeHttpProxy)
+        System.setProperty("http.proxyHost",url.getHost())
+        System.setProperty("http.proxyPort",url.getPort().toString)
+        maybeNoProxy.foreach(System.setProperty("http.nonProxyHosts", _))
+      } catch {
+        case e : MalformedURLException => //Ignore, but should maybe log a warning if the URL is not parsable.
+      }
+    } else {
+      System.clearProperty("http.proxyHost")
+      System.clearProperty("http.proxyPort")
+      if(maybeNoProxy.isEmpty)
+        System.clearProperty("http.nonProxyHosts")
+    }
+
+    val maybeHttpsProxy = state.commonOptions.env.getProperty("https_proxy")
+    if(maybeHttpsProxy != null) {
+      try {
+        val url =  new URL(maybeHttpsProxy)
+        System.setProperty("https.proxyHost",url.getHost())
+        System.setProperty("https.proxyPort",url.getPort().toString)
+        //from java documentation https don't have no_proxy configuration.
+      } catch {
+        case e : MalformedURLException => //Ignore, but should maybe log a warning if the URL is not parsable.
+      }
+    } else {
+      System.clearProperty("https.proxyHost")
+      System.clearProperty("https.proxyPort")
+    }
+
+    val maybeFtpProxy = state.commonOptions.env.getProperty("ftp_proxy")
+    if(maybeFtpProxy != null) {
+      try {
+        val url =  new URL(maybeFtpProxy)
+        System.setProperty("ftp.proxyHost",url.getHost())
+        System.setProperty("ftp.proxyPort",url.getPort().toString)
+        maybeNoProxy.foreach(System.setProperty("ftp.nonProxyHosts", _))
+      } catch {
+        case e : MalformedURLException => //Ignore, but should maybe log a warning if the URL is not parsable.
+      }
+    } else {
+      System.clearProperty("ftp.proxyHost")
+      System.clearProperty("ftp.proxyPort")
+      if(maybeNoProxy.isEmpty)
+        System.clearProperty("http.nonProxyHosts")
+    }
+
+    val maybeSocksProxy = state.commonOptions.env.getProperty("socks_proxy")
+    if(maybeSocksProxy != null) {
+      try {
+        val url =  new URL(maybeSocksProxy)
+        System.setProperty("socksProxyHost",url.getHost())
+        System.setProperty("socksProxyPort",url.getPort().toString)
+      } catch {
+        case e : MalformedURLException => //Ignore, but should maybe log a warning if the URL is not parsable.
+      }
+    } else {
+      System.clearProperty("socksProxyHost")
```

Comment:
I think we should only clear the properties *if* a client environment variable was set before. If, for example, the server was fired up with these properties, and the client doesn't have them, I don't think it's a good behavior to override the proxy properties a server was launched with.
  
This approach should be complementary to the server environment variables, rather than exclusionary. We should also not set the client environment variables if the server proxy settings are also set. We should only set globally client environment variables if the server has not set values for those properties before.
  
In order to achieve this, we need two steps: discovery of these environment variables when we boot up the server and setting the client environment variables based on that information. I would add an `AtomicAny[Set[String]]` field in `Cli` that is initialized the first time we start running `nailMain` with the contents of the server environment variables. Then, I would pass in this set to `setProxyJvmSettings`.

Reply:
I agree for the first part. Fall back on server environement when it's not set by the client.

But I'm not sure about the second. If the user decides to change is proxy environement variable, he would have to restart the server? So settings the proxy per query would be just useful when the server was started without proxy environment variable? In this case the use won't have to restart the server. It seems to me a little bit to "conditional".

Reply:
> But I'm not sure about the second. If the user decides to change is proxy environement variable, he would have to restart the server?
  
I didn't mean that :smile: I think that we should still be able to set client environment variables as many times as they change in the user environment, it's just that we should fast-fail if the server already had them in the first place.

Reply:
A val in Cli object wouldn't be enought to store boot environement variable since it's a singleton object ?

Reply:
And we should load theses environement variable in proxy settings on bootstrap. Because java don't load them automaticaly. ( execept if they are given via -Dhttp.proxyHost ... )

Reply:
> A val in Cli object wouldn't be enought to store boot environement variable since it's a singleton object ?
  
The fact that it's a singleton object is what helps us now that it will contain the environment variables/system properties when the server is initialized. If they are not there when `Cli` was initialized, then we can safely know that the server was not configured to use a proxy.
  
> And we should load theses environement variable in JVM proxy settings on bootstrap. Because Java doesn't load them automatically. ( except if they are given via -Dhttp.proxyHost ... )
  
Sounds good :+1: I was only thinking of system properties but environment variables should be explicitly set during the first step for the rest of the logic to work.